File: bindings/node/e2e/lib/wait.ts
// Copyright 2024 mysql-event-stream Authors
// SPDX-License-Identifier: Apache-2.0

/** Error thrown when a wait operation times out. */
export class WaitTimeoutError extends Error {
  constructor(
    public readonly description: string,
    public readonly timeout: number,
  ) {
    super(`Timed out waiting for ${description} after ${timeout}ms`);
    this.name = "WaitTimeoutError";
  }
}

/**
 * Poll until predicate returns true or timeout is reached.
 */
export async function waitUntil(
  predicate: () => Promise<boolean> | boolean,
  options: {
    timeout?: number;
    interval?: number;
    description?: string;
  } = {},
): Promise<void> {
  const { timeout = 30_000, interval = 1_000, description = "condition" } = options;
  const deadline = Date.now() + timeout;
  let lastError: Error | undefined;

  while (Date.now() < deadline) {
    try {
      if (await predicate()) return;
    } catch (e) {
      lastError = e instanceof Error ? e : new Error(String(e));
    }
    await new Promise((resolve) => setTimeout(resolve, interval));
  }

  const err = new WaitTimeoutError(description, timeout);
  if (lastError) err.cause = lastError;
  throw err;
}

File: bindings/python/e2e/lib/wait.py
"""Polling wait utilities for E2E tests."""

from __future__ import annotations

import time
from collections.abc import Callable


class WaitTimeoutError(Exception):
    """Raised when a wait operation times out."""

    def __init__(self, description: str, timeout: float) -> None:
        super().__init__(f"Timed out waiting for {description} after {timeout}s")
        self.description = description
        self.timeout = timeout


def wait_until(
    predicate: Callable[[], bool],
    *,
    timeout: float = 30.0,
    interval: float = 1.0,
    description: str = "condition",
) -> None:
    """Poll until predicate returns True or timeout is reached.

    Args:
        predicate: Callable that returns True when the condition is met.
        timeout: Maximum seconds to wait.
        interval: Seconds between polls.
        description: Human-readable description for error messages.

    Raises:
        WaitTimeoutError: If the predicate does not return True within the timeout.
    """
    deadline = time.monotonic() + timeout
    last_error: Exception | None = None

    while time.monotonic() < deadline:
        try:
            if predicate():
                return
        except Exception as e:
            last_error = e
        time.sleep(interval)

    msg = f"Timed out waiting for {description} after {timeout}s"
    if last_error:
        msg += f" (last error: {last_error})"
    raise WaitTimeoutError(description, timeout)
